File: src/main/java/com/baulsupp/oksocial/services/surveymonkey/SurveyMonkeyAuthInterceptor.java
package com.baulsupp.oksocial.services.surveymonkey;

import com.baulsupp.oksocial.authenticator.AuthInterceptor;
import com.baulsupp.oksocial.authenticator.JsonCredentialsValidator;
import com.baulsupp.oksocial.authenticator.ValidatedCredentials;
import com.baulsupp.oksocial.completion.ApiCompleter;
import com.baulsupp.oksocial.completion.BaseUrlCompleter;
import com.baulsupp.oksocial.completion.CompletionCache;
import com.baulsupp.oksocial.completion.CompletionQuery;
import com.baulsupp.oksocial.completion.UrlList;
import com.baulsupp.oksocial.credentials.CredentialsStore;
import com.baulsupp.oksocial.credentials.ServiceDefinition;
import com.baulsupp.oksocial.output.OutputHandler;
import com.baulsupp.oksocial.secrets.Secrets;
import java.io.IOException;
import java.util.Collection;
import java.util.List;
import java.util.Optional;
import java.util.concurrent.Future;
import java.util.function.Supplier;
import okhttp3.HttpUrl;
import okhttp3.Interceptor;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;

import static com.baulsupp.oksocial.authenticator.JsonCredentialsValidator.fieldExtractor;

/**
 * https://developer.surveymonkey.com/docs/authentication
 */
public class SurveyMonkeyAuthInterceptor implements AuthInterceptor<SurveyMonkeyToken> {
  @Override public ServiceDefinition<SurveyMonkeyToken> serviceDefinition() {
    return new SurveyMonkeyServiceDefinition();
  }

  @Override public Response intercept(Interceptor.Chain chain, SurveyMonkeyToken credentials)
      throws IOException {
    Request request = chain.request();

    String token = credentials.accessToken;

    HttpUrl newUrl =
        request.url().newBuilder().addQueryParameter("api_key", credentials.apiKey).build();
    request =
        request.newBuilder().addHeader("Authorization", "bearer " + token).url(newUrl).build();

    return chain.proceed(request);
  }

  @Override public SurveyMonkeyToken authorize(OkHttpClient client, OutputHandler outputHandler,
      List<String> authArguments) throws IOException {
    System.err.println("Authorising SurveyMonkey API");

    String apiKey =
        Secrets.prompt("SurveyMonkey API Key", "surveymonkey.apiKey", "", false);
    String accessToken =
        Secrets.prompt("SurveyMonkey Access Token", "surveymonkey.accessToken", "", true);
    return new SurveyMonkeyToken(apiKey, accessToken);
  }

  @Override public Future<Optional<ValidatedCredentials>> validate(OkHttpClient client,
      Request.Builder requestBuilder, SurveyMonkeyToken credentials) throws IOException {
    return new JsonCredentialsValidator(
        SurveyMonkeyUtil.apiRequest("/v3/users/me", requestBuilder),
        fieldExtractor("username")).validate(client);
  }

  @Override public ApiCompleter apiCompleter(String prefix, OkHttpClient client,
      CredentialsStore credentialsStore, CompletionCache completionCache)
      throws IOException {
    Optional<UrlList> urlList = UrlList.fromResource(name());

    Optional<SurveyMonkeyToken> credentials =
        credentialsStore.readDefaultCredentials(serviceDefinition());

    BaseUrlCompleter completer = new BaseUrlCompleter(name(), urlList.get());

    if (credentials.isPresent()) {
      Supplier<Future<List<String>>> supplier = () ->
          CompletionQuery.getIds(client, "https://api.surveymonkey.net/v3/surveys", "data", "id");

      completer.withVariable("survey", supplier, false);
    }

    return completer;
  }

  @Override public Collection<String> hosts() {
    return SurveyMonkeyUtil.API_HOSTS;
  }
}

File: src/main/kotlin/com/baulsupp/oksocial/services/surveymonkey/SurveyMonkeyAuthInterceptor.kt
package com.baulsupp.oksocial.services.surveymonkey

import com.baulsupp.oksocial.authenticator.AuthInterceptor
import com.baulsupp.oksocial.authenticator.ValidatedCredentials
import com.baulsupp.oksocial.authenticator.oauth2.Oauth2ServiceDefinition
import com.baulsupp.oksocial.authenticator.oauth2.Oauth2Token
import com.baulsupp.oksocial.completion.ApiCompleter
import com.baulsupp.oksocial.completion.BaseUrlCompleter
import com.baulsupp.oksocial.completion.CompletionVariableCache
import com.baulsupp.oksocial.completion.UrlList
import com.baulsupp.oksocial.credentials.CredentialsStore
import com.baulsupp.oksocial.credentials.Token
import com.baulsupp.oksocial.credentials.TokenValue
import com.baulsupp.oksocial.kotlin.query
import com.baulsupp.oksocial.output.OutputHandler
import com.baulsupp.oksocial.secrets.Secrets
import com.baulsupp.oksocial.services.surveymonkey.model.SurveyList
import okhttp3.Interceptor
import okhttp3.OkHttpClient
import okhttp3.Response

/**
 * https://developer.surveymonkey.com/docs/authentication
 */
class SurveyMonkeyAuthInterceptor : AuthInterceptor<Oauth2Token>() {
  override fun serviceDefinition(): Oauth2ServiceDefinition {
    return Oauth2ServiceDefinition("api.surveymonkey.net", "Survey Monkey API", "surveymonkey",
      "https://developer.surveymonkey.com/api/v3/#scopes",
      "https://developer.surveymonkey.com/apps/")
  }

  override fun intercept(chain: Interceptor.Chain, credentials: Oauth2Token): Response {
    val newRequest = chain.request().newBuilder().addHeader("Authorization",
      "bearer " + credentials.accessToken).build()

    return chain.proceed(newRequest)
  }

  override suspend fun authorize(
    client: OkHttpClient,
    outputHandler: OutputHandler<Response>,
    authArguments: List<String>
  ): Oauth2Token {

    val clientId = Secrets.prompt("SurveyMonkey Client ID", "surveymonkey.clientId", "", false)
    val clientSecret = Secrets.prompt("SurveyMonkey Client Secret", "surveymonkey.clientSecret", "",
      true)
    return SurveyMonkeyAuthFlow.login(client, outputHandler, clientId, clientSecret)
  }

  data class User(
    val username: String,
    val first_name: String?,
    val last_name: String?,
    val email: String
  )

  override suspend fun validate(
    client: OkHttpClient,
    credentials: Oauth2Token
  ): ValidatedCredentials {
    val user = client.query<User>("https://api.surveymonkey.net/v3/users/me",
      TokenValue(credentials))

    return if (user.first_name != null && user.last_name != null) {
      ValidatedCredentials(user.first_name + " " + user.last_name)
    } else {
      ValidatedCredentials(user.email)
    }
  }

  override fun apiCompleter(
    prefix: String,
    client: OkHttpClient,
    credentialsStore: CredentialsStore,
    completionVariableCache: CompletionVariableCache,
    tokenSet: Token
  ): ApiCompleter {
    val urlList = UrlList.fromResource(name())

    val completer = BaseUrlCompleter(urlList!!, hosts(), completionVariableCache)

    completer.withCachedVariable(name(), "survey", {
      credentialsStore.get(serviceDefinition(), tokenSet)?.let {
        client.query<SurveyList>("https://api.surveymonkey.net/v3/surveys", tokenSet).data.map { m -> m.id }
      }
    })

    return completer
  }

  override fun hosts(): Set<String> = setOf("api.surveymonkey.net")
}
